### PR TITLE
629 OpenMP parallelization of ABM

### LINKED ISSUE #629: Parallelize ABM
### Feature description

the ABM should run in parallel using OpenMP.

Requires an improved random number generator.

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### PR DESCRIPTION
**Pull request by dabele**
## Changes

- Parallelized ABM using OpenMP
- Iterate over Persons in parallel during interaction/migration.
- Requires some synchronization of locations.
- Iterate over Locations in parallel during begin/end_step.
- New counter based random number generator
- persons use their own RNG for reproduceability independent of number of threads and to avoid synchronizating the global RNG
- enabled warnings and warnings-as-errors for all examples to get rid of build output spam.

Closes #629

## Merge Request - GuideLine Checklist

- *Check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow) before opening a Pull request/Merge request.*
- *Request a reviewer when your work is ready to review, before this please use the draft feature.*

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

### Additional context

Please list additional information or things a reviewer should look out for.

### CODE REVIEW

**Comment by dabele** [comment]
scaling results using the ABM simulation with different numbers of days simulated and different numbers of OMP threads

```
10 days

1 34
2 19
4 12
8 11

30 days

1 99
2 57
4 34
8 21

60 days

1 206
2 113
4 64
8 38
```

Scaling improves with the length of the simulation.

Generally scaling could be better, but it's probably good enough as a first version.

**Comment by dabele** [comment]
I checked to see if the results are reproducible no matter the number of OMP threads used (with fixed RNG seeds of course). It works except for one issue.

During migration, the capacity of locations is checked and access denied. During a time step, some persons are allowed to migrate, later persons are denied. But the sequence of persons is random (or at least different) when more threads are used, because the persons are distributed to the threads. 

If I comment out the capacity check, the result is always the same, independent of the number of threads, so as far as I can tell it's the only issue preventing full reproducibility. 

Not sure how to handle this. This could be considered a problem even without multithreading. If you change the order of persons in the world by adding them in a different sequence, the result changes because a different person comes first and gets to migrate to the location. (Note that if you change the order of persons, the result might change anyway because random numbers are generated in a different order)

Does this need to be resolved in this issue? If so, how?

Maybe ignore capacity during a time step by caching the current number of persons at the beginning of a time step? This could mean that some locations are filled beyond capacity if a lot of persons migrate there in the same timestep.

Maybe store migrating persons in a different list and at the end of a timestep do some kind of deterministic conflict resolution to decide who gets to migrate. Would probably make scaling a bit worse because it moves a bit of logic out of the OpenMP loop, but shouldn't be a lot worse.

Maybe another idea. Do we need a meeting to discuss this?

**Comment by dabele** [comment]
Need to fix the merge conflicts, but those shouldn't affect the general design, so please review anyway.

**Comment by mknaranja** [comment]
Thank you @dabele 

Maybe just two comments on your posts before I have done the review:

1) Capacity:
a) I think this is an important issue, however, it could maybe be addressed in a separate issue and for the moment we restrict repoducibility to the same number of threads or deactivated capacity. One can still close the whole location and the result is reproducible. 
b) After a brief consideration of the situation it seems to me that the location needs to know where comes to that location. However, this still does not resolve the problem as people can arrive at different times. We probably need another meeting.

2) I was wondering why it scales (and should scale) with the number of days. This would be parallel in time, isn't it? I would have thought it to scale with the number of persons.

**Comment by dabele** [comment]
> I was wondering why it scales (and should scale) with the number of days. This would be parallel in time, isn't it? I would have thought it to scale with the number of persons.

It will probably scale better with more persons. The reason I tried more days was to see the impact of pre and post processing, which is mostly constant runtime.

File: cpp/tests/test_random_number_generator.cpp [hunk 1]
```diff
@@ -4,35 +4,43 @@
 #include <algorithm>
 #include <iterator>
 
-TEST(RandomNumberGenerator, blocks)
```
Comment (mknaranja) [inline cpp/tests/test_random_number_generator.cpp:7]: There is a test on the subsequence_counter functioning in persons. Should we test the helper functions in `cpp/memilio/utils/random_number_generator.h` more? I can't see the code coverage right now, as there are merge conflicts. Maybe check after resolving.

File: cpp/examples/CMakeLists.txt [hunk 2]
```diff
@@ -4,57 +4,73 @@ configure_file(data_dir.h.in data_dir.h)
 
 add_executable(euler_example euler_test.cpp)
 target_link_libraries(euler_example PRIVATE memilio)
+target_compile_options(euler_example PRIVATE ${MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS})
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/examples/CMakeLists.txt:7]: the warning compiler flags that we set in memilio are deliberately not public. so they need to be enabled in each target seperately. This was just never done for the examples, so there were quite a few warnings in them, some of them spamming the build output. Has nothing to do with Random123, just made building the project more annoying while I tried to get scorep to work, so I fixed it here.

File: cpp/examples/abm_minimal.cpp [hunk 3]
```diff
@@ -98,8 +99,9 @@ int main()
     for (auto& person : persons) {
         mio::abm::InfectionState infection_state =
             (mio::abm::InfectionState)(rand() % ((uint32_t)mio::abm::InfectionState::Count - 1));
+        auto rng = mio::abm::Person::RandomNumberGenerator(world.get_rng(), person);
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/examples/abm_minimal.cpp:102]: not copying, always creating a new Rng for the person on demand. the personal rng is not persistent, only the rng counter is persistent.

File: cpp/thirdparty/CMakeLists.txt [hunk 4]
```diff
@@ -112,3 +113,25 @@ endif()
 if (MEMILIO_ENABLE_MPI)
     find_package(MPI REQUIRED COMPONENTS CXX)
 endif()
+
+if (MEMILIO_ENABLE_OPENMP)
+    find_package(OpenMP COMPONENTS CXX REQUIRED)
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/thirdparty/CMakeLists.txt:118]: no difference, it's like command line flags

File: cpp/memilio/utils/random_number_generator.h [hunk 5]
```diff
@@ -20,81 +20,231 @@
...
+    static_assert(sizeof(U) > sizeof(typename C::ValueType), "Subsequence counter must be smaller than requested total counter.");
+    //TODO allow sizeof(C) = sizeof(U) and check actual values?
+    //TODO: assert value of subequence_idx
+
+    //
+    //high bits: subsequence idx
+    //low bits: subsequence counter
+    static const U bytes_shift = sizeof(U) - sizeof(C);
+    auto s                     = static_cast<U>(subsequence_idx);
+    return Counter<U>{(s << (bytes_shift * 8)) + static_cast<U>(counter.get())};
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:149]: 8 for number of bits per byte. I'll give it a variable name.

File: cpp/memilio/utils/random_number_generator.h [hunk 6]
```diff
@@ -269,7 +396,7 @@ class DistributionAdapter
      */
     static DistributionAdapter& get_instance()
     {
-        static thread_local DistributionAdapter instance;
+        static DistributionAdapter instance;
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:399]: need to check again

File: cpp/models/abm/lockdown_rules.cpp [hunk 7]
```diff
@@ -30,18 +31,21 @@ void set_home_office(TimePoint t_begin, double p, MigrationParameters& params)
 {
     auto damping1 = Eigen::VectorXd::Constant(1, p);
     params.get<WorkRatio>().add_damping(damping1, SimulationTime(t_begin.days()));
+    params.get<WorkRatio>().finalize();
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/lockdown_rules.cpp:34]: maybe we can gather all the finalize calls somewhere to make it clear when this should be called (outside of parallel regions).

But we can rethink the finalize in general. Quick examplanation of finalize: Dampings do some computation (adding up all the dampings) and cache the result. The cache is created on demand (e.g., by calling `get_matrix_at`, not thread safe) or by calling finalize. So we call finalize after adding dampings to prepare the cache before entering a parallel region, then `get_matrix_at` just reads the existing cache, which is thread safe. 

I don't think giving every damping a mutex to snychronize the cache so finalize is a good idea, mutexes are heavy and difficult to move around, dampings are a basic building block and should be moveable/copieable. But we might make finalize unnecessary in most cases by switching the defaults. In most cases (like in the example here), immediately recomputing the cache after adding dampings is fine. But in some cases this is expensive (e.g., dynamic NPIs, which could add quite a lot of dampings at many time steps during the simulation). We could make it so the cache needs to be explicitly disabled (and enabled after adding dampings), otherwise the cache is immediatly recomputed on every `add_damping`. Then the default is for safety. Right now the default is for performance.

File: cpp/models/abm/simulation.cpp [hunk 8]
```diff
@@ -18,6 +18,9 @@
 * limitations under the License.
 */
 #include "abm/simulation.h"
+#include "memilio/utils/logging.h"
+#include "memilio/utils/random_number_generator.h"
+#include <random>
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/simulation.cpp:23]: I wrapped the #pragmas in preprocessor macros so they disappear when OpenMP is disabled to avoid this warning. Probably missed a #pragma somewhere. I think we talked about this.

File: cpp/models/abm/simulation.cpp [hunk 9]
```diff
@@ -47,20 +50,32 @@ void Simulation::advance(TimePoint tmax)
...
     m_result.add_time_point(t.days());
     m_result.get_last_value().setZero();
-    for (auto&& locations : m_world.get_locations()) {
-        for (auto&& location : locations) {
-            m_result.get_last_value() += location->get_subpopulations().get_last_value().cast<ScalarType>();
+    #pragma omp parallel 
+    {
+        Eigen::VectorXd sum = Eigen::VectorXd::Zero(m_result.get_num_elements());
+        for (auto&& locations : m_world.get_locations()) {
+            #pragma omp for
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/simulation.cpp:70]: I'll add a comment here. This is basically a parallel reduction, but the `reduction(op:+)` clause of openmp doesn't work well for `Eigen::Vector`. So this just a manual reduction, make one sum per thread, then with a mutex add up the local sums at the end.

File: cpp/models/abm/world.cpp [hunk 10]
```diff
@@ -53,22 +55,29 @@ Person& World::add_person(const LocationId id, AgeGroup age)
...
 }
 
 void World::interaction(TimePoint t, TimeSpan dt)
 {
-    for (auto&& person : m_persons) {
-        person->interact(t, dt, m_infection_parameters);
+    PRAGMA_OMP(parallel for)
+    for (auto i = size_t(0); i < m_persons.size(); ++i) {
+        auto&& person = m_persons[i];
+        auto personal_rng = Person::RandomNumberGenerator(m_rng, *person);
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/world.cpp:70]: Counter based RNGs are cheap to make/recreate if you store the counter somewhere. We want all RNGs in the simulation to share the seeded key, but we don't want to store it everywhere (memory size and consistency). So we pass the key around as function argument, store the counter per person and create an actual RNG (an object with `operator()` that has the key and increments the counter) on demand.

I'll try to find a good place to document this.

File: cpp/memilio/utils/random_number_generator.h [hunk 11]
```diff
@@ -20,81 +20,231 @@
...
+* @param subsequence_idx The index of the subsequence.
+* @param counter The counter in the subsequence.
+* @param return The counter in the total sequence.
+*/
+template <class U, class T, class C>
+Counter<U> rng_subsequence_counter(T subsequence_idx, C counter)
+{
+    static_assert(sizeof(U) > sizeof(typename C::ValueType), "Subsequence counter must be smaller than requested total counter.");
+    //TODO allow sizeof(C) = sizeof(U) and check actual values?
+    //TODO: assert value of subequence_idx
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:142]: basically, sequence index and counter together should not overflow uint64. I think this can be added.

File: cpp/memilio/utils/random_number_generator.h [hunk 12]
```diff
@@ -108,46 +258,16 @@ class RandomNumberGenerator
             m_seeds.assign(num_seeds, 0);
         }
         MPI_Bcast(m_seeds.data(), num_seeds, MPI_UNSIGNED, 0, mpi::get_world());
+        if (rank != 0) {
+            seed(m_seeds);
+        }
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:263]: Possibly. I can't be sure, but maybe I only tested the MPI parallelization with a fixed seed.

File: cpp/tests/test_abm_location.cpp [hunk 13]
```diff
@@ -85,50 +87,52 @@ TEST(TestLocation, CacheExposureRate)
...
-        location.cache_exposure_rates(t, dt);
-
-        EXPECT_NEAR((location.get_cells()[0].m_cached_exposure_rate_air[{variant}]), 0.0061363631046134817, 1e-14);
-        EXPECT_NEAR((location.get_cells()[1].m_cached_exposure_rate_air[{variant}]), 0.0030681815523067408, 1e-14);
-        EXPECT_NEAR((location.get_cells()[2].m_cached_exposure_rate_air[{variant}]), 0, 1e-14);
-    }
+    auto rng = mio::RandomNumberGenerator();
+
+    mio::abm::AgeGroup age = mio::abm::AgeGroup(
+        mio::UniformIntDistribution<int>::get_instance()(rng, 0, int(mio::abm::AgeGroup::Count) - 1));
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/tests/test_abm_location.cpp:93]: There is no mocking here, we are only using the RNG to randomize parameters that SHOULD have no influence on the test. Because there is definitely no mocks in place, we could just use the constructor to create a new distribution object instead of using the distribution from `get_instance`. But now I made the constructor of the distributions private for safety, because we want to make sure that `get_instance` is used during the simulation to enable mocking. Because the constructor is private, we need to use `get_instance` here as well, even though it's not necessary for mocking.

File: cpp/models/abm/migration_rules.cpp [hunk 14]
```diff
@@ -33,22 +33,24 @@ namespace mio
 namespace abm
 {
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/migration_rules.cpp:35]: why? `Person` is still used everywhere. I added RNG as the first argument of the migration rules, but did not remove any argument.

File: cpp/tests/test_parameter_studies.cpp [hunk 15]
```diff
@@ -154,6 +154,7 @@ TEST(ParameterStudies, sample_graph)
     graph.add_edge(0, 1, mio::MigrationParameters(Eigen::VectorXd::Constant(Eigen::Index(num_groups * 8), 1.0)));
 
     auto study   = mio::ParameterStudy<mio::osecir::Simulation<>>(graph, 0.0, 0.0, 0.5, 1);
+    mio::log_rng_seeds(study.get_rng(), mio::LogLevel::warn);
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/tests/test_parameter_studies.cpp:157]: added doc string

File: cpp/tests/test_random_number_generator.cpp (hunk 1 above)
Comment (dabele) [reply cpp/tests/test_random_number_generator.cpp:7]: added more tests for the counter based RNG stuff

File: cpp/models/abm/infection.h [hunk 16]
```diff
@@ -49,13 +50,16 @@ class Infection
 public:
     /**
      * @brief Create an Infection for a single Person.
+     * Draws a random infection course.
+     * @param[inout] rng Random number generator.
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/infection.h:54]: done

File: cpp/memilio/utils/random_number_generator.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:149]: rewrote the function to be more clear

File: cpp/models/abm/lockdown_rules.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/lockdown_rules.cpp:34]: Maybe a new issue?

File: cpp/memilio/utils/random_number_generator.h [hunk 17]
```diff
@@ -20,81 +20,231 @@
...
+inline r123array2x32 to_r123_array(uint64_t i)
+{
+    threefry2x32_ctr_t c;
+    std::memcpy(c.data(), &i, sizeof(uint64_t));
+    return c;
+}
+}
+
+/**
+* A key type for counter based random number generators.
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:71]: added extensive documentation of counter based random number generators.

File: cpp/models/abm/person.h [hunk 18]
```diff
@@ -48,13 +49,74 @@ static constexpr uint32_t INVALID_PERSON_ID = std::numeric_limits<uint32_t>::max
...
+    {
+    public:
+        /**
+        * Creates a random number generator for a person.
+        * @param key Key to be used by the generator.
+        * @param person Person who's counter will be used. 
+        */
+        RandomNumberGenerator(Key<uint64_t> key, Person& person)
+            : m_person(person)
+            , m_key(key)
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/person.h:67]: changed

File: cpp/memilio/compartments/parameter_studies.h [hunk 19]
```diff
@@ -148,19 +150,25 @@ class ParameterStudy
...
+            //assume that sampling the graph uses the thread local rng and isn't multithreaded
+            auto initial_rng_counter =
+                rng_subsequence_counter<uint64_t>(static_cast<uint32_t>(run_idx), Counter<uint32_t>(0));
+            m_rng.set_counter(initial_rng_counter);
+            thread_local_rng() = m_rng;
 
+            //sample
             auto sim = create_sampled_simulation(sample_graph);
+
+            m_rng = thread_local_rng();
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/compartments/parameter_studies.h:165]: added comments and fixed the code.

**Comment by dabele** [comment]
I think I adressed all the issues. 

I also resolved merge conflicts. maybe @DavidKerkmann or someone who worked on ABM recently can check I didn't break any of the recent changes.

I had to fix two weird bugs to get python bindings and MSVC build running.

MSVC doesn't do emtpy base optimization correctly, which we rely on for our TypeSafe class; a TypeSafe<int> should be the size of an, but on MSVC it's bigger when adding operator+ using the OperatorAdd, etc base classes. I hope I explained the fix in the code comments well enough, it's pretty safe anyway, the worst thing to happen if the types are a bit larger is slightly degraded performance. See MEMILIO_ENABLE_EBO macro.

Pybind11 thinks that our Range class is a real container because it has `Range::value_type` and iterators and everything. So it checks the value_type to see if `Range` can be copied. But `Range`  is just a view, so copying the Range doesn't copy the elements. The fix is a bit annoying, but again I hope the documentation is sufficient. See PYMIO_IGNORE_VALUE_TYPE macro.

**Comment by dabele** [comment]
SInce I did some performance analysis with Scorep, I wanted to put down somewhere the steps I took in case this is needed in the future.

What you need:
* Scorep for instrumentation
* Cube for visualization
* Scalasca for traces (more precise than just sampling)
How to get them depends on your system, not writing it here, but they should be available on most clusters through the module system.

to build (requires scorep):
```
export SCOREP_WRAPPER=off #disable scorep during cmake configure
cmake .. -DMEMILIO_ENABLE_OPENMP=ON -DCMAKE_CXX_COMPILER=${Path to scorep}/scorep-mpic++ -DCMAKE_C_COMPILER=${Path to scorep}/scorep-mpicc -DCMAKE_BUILD_TYPE=RelWithDebInfo
export SCOREP_WRAPPER=on #reenable scorep
export SCOREP_WRAPPER_INSTRUMENTER_FLAGS="--pomp --thread=omp --mpp=none --instrument-filter=./filter-main"
cmake --build . -j --target=abm_simulation
```
I'm not sure about every one of these instrumenter flags, but it builds that way, and openmp instrumentation is enabled. there are corresponding flags for MPI.

The filter file is basically always necessary to get a reasonable result. Otherwise you get millions of results for deep template helper functions. I attached the filter file that was most useful for the ABM, but that may have to be adapted further.

You may get a lot of build errors about "include style" and "gcc extension". These are enabled by the `-pedantic` flag that we set in CMake. You need to comment out the lines that enable compiler warnings in the CMakeLists.txt.

It is useful to use "--target" to set the target to build, because scorep instrumentation has difficulties with some of the other targets, e.g. the 2021_vaccination_sarscov2_delta_germany. Also, due to instrumentation there will be a full rebuild for every source change, so building only the target you need speeds up development a lot.

To run the basic analysis with sampling:
```
export OMP_NUM_THREADS=8 #or whatever
./bin/abm_simulation <some result dir> 1
```
There will be a directory named `scorep_...` with a `.cubex` file. That can  be opened in the Cube visualization tool. This can give a first indication for bottlenecks, then you don't need Scalasca.

For deeper analysis, use Scalasca for tracing:
```
export OMP_NUM_THREADS=8 #or whatever
scalasca -analyze -q -t ./bin/abm_simulation <result dir> 1
scalasca -examine scorep_abm_simulation_Ox8_trace
```
The first command runs the program and creates the trace files.
The second command opens Cube to visualize the traces, this time with better information about load imbalances etc.

**Comment by DavidKerkmann** [comment]
> I checked to see if the results are reproducible no matter the number of OMP threads used (with fixed RNG seeds of course). It works except for one issue.
> 
> During migration, the capacity of locations is checked and access denied. During a time step, some persons are allowed to migrate, later persons are denied. But the sequence of persons is random (or at least different) when more threads are used, because the persons are distributed to the threads.
> 
> If I comment out the capacity check, the result is always the same, independent of the number of threads, so as far as I can tell it's the only issue preventing full reproducibility.
> 
> Not sure how to handle this. This could be considered a problem even without multithreading. If you change the order of persons in the world by adding them in a different sequence, the result changes because a different person comes first and gets to migrate to the location. (Note that if you change the order of persons, the result might change anyway because random numbers are generated in a different order)
> 
> Does this need to be resolved in this issue? If so, how?
> 
> Maybe ignore capacity during a time step by caching the current number of persons at the beginning of a time step? This could mean that some locations are filled beyond capacity if a lot of persons migrate there in the same timestep.
> 
> Maybe store migrating persons in a different list and at the end of a timestep do some kind of deterministic conflict resolution to decide who gets to migrate. Would probably make scaling a bit worse because it moves a bit of logic out of the OpenMP loop, but shouldn't be a lot worse.
> 
> Maybe another idea. Do we need a meeting to discuss this?

We discussed this again and decided to the following solution for now:
In the parallel implementation, the check for capacity is disabled. We add all persons to the location and do some conflict resolution after the migration step as you suggested. This can be done via an additional buffer that first saves all persons that are new to the location or we can simply use the `m_time_at_location` variable to check for all new persons. We then throw out all people that don't fit to a fixed criterion. As a first step, this can just be their ID, so the highest IDs have to leave. This makes it reproducible as long as the persons are not created in parallel so that their ID is always the same. Also, the people who have to leave could either stay at their previous location or go home instead. Currently, they would just stay but we decided that it would be better to send them home in this case (this would also impact the non-parallel implementation).
This conflict resolution might even be an enhancement to the code since there might be people being blocked to go to a location even though other people are leaving at the same time point (but are being processed later in the code). This can also happen when executing with only one core, as you also mentioned, so it might be a good idea to always do this and always remove capacity check.

TL;DR:
1. Remove capacity check (for parallel code or in general).
2. Conflict resolution: Check for person ID in locations where capacity is exceeded and send new people with highest IDs home.

**Comment by mknaranja** [comment]
> This can also happen when executing with only one core, as you also mentioned, so it might be a good idea to always do this and always remove capacity check.

Thanks @DavidKerkmann. This is a very good summary and you even raised an important point we forgot in our meeting. I opt for going with the solution described by David for now.

**Comment by dabele** [comment]
From a modelling point this sounds good. Need to see how this is best implemented with some of the performance optimizations I have been working on.

File: cpp/thirdparty/CMakeLists.txt [hunk 20]
```diff
@@ -112,3 +113,25 @@ endif()
...
+
+if (MEMILIO_ENABLE_OPENMP)
+    find_package(OpenMP COMPONENTS CXX REQUIRED)
+endif()
+
+#Random123 library for random number generators
+message(STATUS "Downloading Random123 library")
+
+include(FetchContent)
+FetchContent_Declare(Random123
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/thirdparty/CMakeLists.txt:125]: added to wiki

File: cpp/models/abm/lockdown_rules.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/lockdown_rules.cpp:34]: #753

**Review by DavidKerkmann — COMMENTED** [review]
Thanks for putting this together. I have only a couple of questions and would like to discuss how to proceed further with this. There have been some changes to the code in the meantime, but nothing too major to not be included rather easily, I think.

File: cpp/memilio/utils/random_number_generator.h [hunk 21]
```diff
@@ -17,84 +17,323 @@
...
+/**
+* A counter type for counter based random number generators.
+* @tparam an unsigned integer type that determines the size of the counter, i.e., the length of the random sequence.
+*/
+template <class T>
+struct MEMILIO_ENABLE_EBO Counter : TypeSafe<T, Counter<T>>, OperatorComparison<Counter<T>>, OperatorAdditionSubtraction<Counter<T>> {
+    static_assert(std::is_unsigned<T>::value, "Underlying Integer type must be unsigned.");
+    using TypeSafe<T, Counter<T>>::TypeSafe;
+};
+static_assert(sizeof(Counter<uint32_t>) == sizeof(uint32_t), "");
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:188]: didn't think to do it because it's not that important, since we only store the key once.

File: cpp/models/abm/person.h [hunk 22]
```diff
@@ -48,13 +49,83 @@ static constexpr uint32_t INVALID_PERSON_ID = std::numeric_limits<uint32_t>::max
...
+    */
+    class RandomNumberGenerator : public RandomNumberGeneratorBase<RandomNumberGenerator>
+    {
+    public:
+        /**
+        * Creates a RandomNumberGenerator for a person.
+        * @param key Key to be used by the generator.
+        * @param person Person who's counter will be used. 
+        */
+        RandomNumberGenerator(Key<uint64_t> key, Person& person)
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/person.h:74]: To allow inlining. Ideally the RNG is completely optimized away, but I haven't checked that it is.

File: cpp/models/abm/simulation.h [hunk 23]
```diff
@@ -22,6 +22,7 @@
 
 #include "abm/world.h"
 #include "abm/time.h"
+#include "memilio/utils/random_number_generator.h"
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/simulation.h:25]: At some point the simulation held the RNG. I think this is a remnant from that.

File: cpp/models/abm/world.cpp [hunk 24]
```diff
@@ -52,24 +54,28 @@ Person& World::add_person(const LocationId id, AgeGroup age)
...
     migration(t, dt);
     end_step(t, dt);
 }
 
 void World::interaction(TimePoint t, TimeSpan dt)
 {
-    for (auto&& person : m_persons) {
-        person->interact(t, dt, m_infection_parameters);
+    PRAGMA_OMP(parallel for)
+    for (auto i = size_t(0); i < m_persons.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/world.cpp:67]: openmp can't do range for loops

File: cpp/memilio/utils/random_number_generator.h [hunk 25]
```diff
@@ -17,84 +17,323 @@
...
+        m_counter = counter;
+    }
+    void increment_counter()
+    {
+        ++m_counter;
+    }
+    static std::vector<uint32_t> generate_seeds()
+    {
+        std::random_device rd;
+        return {rd(), rd(), rd(), rd(), rd(), rd()};
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h:318]: the number is arbitrary. you want at least as 2 (64 bit) to use the whole range of the 64 bit key. I used 6 for mersenne twister, those RNGs can store a few thousand bits of randomness. I guess 6 is a bit overkill but doesn't hurt.

File: cpp/tests/test_abm_world.cpp [hunk 26]
```diff
@@ -351,6 +353,8 @@ TEST(TestWorld, evolveMigration)
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/tests/test_abm_world.cpp:353]: Yes, but I don't think this should be it's own unit test. Instead we should run the normal unittests (or a subset) with OMP_NUM_THREADS > 1. This requires a bit of setup, since normal github runners don't really support multithreading.

File: cpp/models/abm/simulation.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/models/abm/simulation.h]: removed obsolete includes

File: cpp/memilio/utils/random_number_generator.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/random_number_generator.h]: added an assert for Key

File: cpp/tests/test_abm_world.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/tests/test_abm_world.cpp:346]: Created an issue for multi threaded tests: #786

**Comment by dabele** [comment]
Finished Code Review revisions.

Merged with main branch to resolve merge conflicts.

Added a constructor to Person::RandomNumberGenerator that does not require an actual Person instance, so that the RNG is easier to create in Unittests.

**Review by DavidKerkmann — APPROVED** [review]
Thanks for the latest changes, the inclusion of the new constructor for the Person rng looks very useful. I have no further change requests.

**Review by mknaranja — APPROVED** [review]
